fix usage of `autodiff` macro with inner functions

haenoe opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

This PR adds additional handling into the expansion step of the `std::autodiff` macro (in `compiler/rustc_builtin_macros/src/autodiff.rs`), which allows the macro to be applied to inner functions.

```rust
#![feature(autodiff)]
use std::autodiff::autodiff;

fn main() {
    #[autodiff(d_inner, Forward, Dual, DualOnly)]
    fn inner(x: f32) -> f32 {
        x * x
    }
}
```

Previously, the compiler didn't allow this due to only handling `Annotatable::Item` and `Annotatable::AssocItem` and missing the handling of `Annotatable::Stmt`. This resulted in the rather generic error

```
error: autodiff must be applied to function
 --> src/main.rs:6:5
  |
6 | /     fn inner(x: f32) -> f32 {
7 | |         x * x
8 | |     }
  | |_____^

error: could not compile `enzyme-test` (bin "enzyme-test") due to 1 previous error
```

This issue was originally reported [here](https://github.com/EnzymeAD/rust/issues/184).

Quick question: would it make sense to add a ui test to ensure there is no regression on this?
This is my first contribution, so I'm extra grateful for any piece of feedback!! :D

r? @oli-obk 

Tracking issue for autodiff: #124509

haenoe committed:
fix usage of `autodiff` macro with inner functions

- fix errors caused by the move of `ast::Item::ident` (see #138740)
- move the logic of getting `sig`, `vis`, and `ident` from two seperate
  `match` statements into one (less repetition especially with the
  nested `match`)

ZuseZ4 commented:
Thank you for looking into this! Yes, adding a test to avoid regressions would be great.
UI tests are usually more towards error testing, since this one now works I would add it under `tests/pretty/autodiff_reverse.{rs|pp}`. You can just add an example by renaming main to f6, and use `--bless` when running the test.
With `git diff` you can then check if the change is what you expect it to be.
The whole test suite just got reworked recently, so it has quite nice docs these days: https://rustc-dev-guide.rust-lang.org/tests/compiletest.html
https://rustc-dev-guide.rust-lang.org/tests/intro.html
and https://rustc-dev-guide.rust-lang.org/tests/best-practices.html are probably the relevant ones.

bors commented:
:v: @ZuseZ4, you can now approve this pull request!

If @oli-obk told you to "`r=me`" after making some further change, please make that change, then do `@bors r=@oli-obk`
<!-- homu: {"type":"Delegated","delegator":"oli-obk","delegate":"ZuseZ4"} -->

haenoe commented:
> Thank you for looking into this! Yes, adding a test to avoid regressions would be great. UI tests are usually more towards stdout/stderr (or runtime) testing. Since this one now works, I would add it under `tests/pretty/autodiff_reverse.{rs|pp}`. You can just add an example by renaming main to f6, and use `--bless` when running the test. With `git diff` you can then check if the change is what you expect it to be. The whole test suite just got reworked recently, so it has quite nice docs these days: https://rustc-dev-guide.rust-lang.org/tests/compiletest.html https://rustc-dev-guide.rust-lang.org/tests/intro.html and https://rustc-dev-guide.rust-lang.org/tests/best-practices.html are probably the relevant ones.

@ZuseZ4 , thank you for providing more insights into this! The pretty printer tests do indeed sound like a good place for this. Also I've got to express how cool and insightful the docs are -- this is really not something to be taken as granted :D

However, while looking into this I noticed that both `tests/pretty/autodiff_{forward,reverse}.rs` are currently failing due to the fact that they still include functions which use the `autodiff` macro in conjuction with `DiffMode::{Forward,Reverse}First`. These two were removed with commit [061abbc36928cce784c54463c266f4d43d14d419](https://github.com/rust-lang/rust/commit/061abbc36928cce784c54463c266f4d43d14d419).

It makes sense to open a separate PR fixing these issues before continuing with the inner function test, right?
Have a nice evening ^^'

ZuseZ4 commented:
Yes I agree, I think one reason why I ended up being a rustc developer rather than working on other projects is the documentation. I don't have a pc this weekend, but my batching PR contains fixes for those. You can cherry pick them if you want, or I'll make a PR for that on Monday. I had hoped to get it done for the weekend, but unfortunately ran out of time.

haenoe commented:
Okidoki! Don't stress, I'll just rebase when the changes are in master.
Enjoy your weekend!

bors commented:
:umbrella: The latest upstream changes (presumably #138791) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

haenoe committed:
feat: add test to validate autodiff macro expansion

haenoe commented:
Hey @ZuseZ4 
Rebased this to master and also tried to integrate your fixes regarding the duplicated attributes via your `same_attribute` function.
Quick question: Do you think this simple test suffices to test the expansion logic? Or would you add one/two more advanced usages (e.g. multiple derivatives).

ZuseZ4 commented:
looks good. I would just add a second autodiff macro with an in any way different config to the same function/test, to verify that even after future refactors we don't add inline or autodiff attributes multiple time to the source functions.

bors commented:
:umbrella: The latest upstream changes (presumably #138933) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

haenoe committed:
feat: apply autodiff macro twice to inner function

Verify that the expanded `inline` and `rustc_autodiff` macros are not
duplicated.

haenoe commented:
Hey @ZuseZ4!
Sorry for the long silence, I didn't have access to my computer :/
But now I am able to be more responsive, so let me know if you have any more feedback!

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14187967980/job/39746624064) [(plain)](https://github.com/rust-lang/rust/commit/013edbe0899c440701af917d8c4cde943da43dc5/checks/39746624064/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 19.2s done
#19 DONE 48.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error: cannot find macro `bug` in this scope
   --> compiler/rustc_builtin_macros/src/autodiff.rs:390:17
    |
390 |                 bug!("item kind checked previously")
    |                 ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_builtin_macros/src/autodiff.rs:346:17
    |
346 |                 bug!("annotatable kind checked previously")
    |                 ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_builtin_macros/src/autodiff.rs:340:26
    |
340 |                     _ => bug!("stmt kind checked previously"),
    |                          ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_builtin_macros/src/autodiff.rs:211:22
    |
211 |                 _ => bug!("stmt kind checked previously"),
    |                      ^^^

error[E0423]: expected value, found variant `Impl`
   --> compiler/rustc_builtin_macros/src/autodiff.rs:369:46
    |
369 |                 Annotatable::AssocItem(d_fn, Impl)
    |                                              ^^^^
    |
   ::: /checkout/compiler/rustc_ast/src/visit.rs:27:5
    |
27  |     Impl { of_trait: bool },
    |     ---- `Impl` defined here
    |
help: use struct literal syntax instead
    |
369 |                 Annotatable::AssocItem(d_fn, Impl { of_trait: val })
    |                                                   +++++++++++++++++
help: consider importing one of these items instead
    |
7   +     use crate::autodiff::llvm_enzyme::ItemKind::Impl;
    |
7   +     use crate::autodiff::llvm_enzyme::ast::visit::BoundKind::Impl;
    |
7   +     use crate::autodiff::llvm_enzyme::kw::Impl;
    |
7   +     use rustc_ast::ItemKind::Impl;
    |
      and 5 other candidates

```

</details>

ZuseZ4 commented:
No worries, and the test looks good, thanks!
However something else seem to have broken in the meantime, maybe you can try to rebase locally onto the latest master and see if that builds?
Also, https://github.com/rust-lang/rust/pull/138740 also modifies a lot of code, including our frontend, so you'll likely need to rebase once it lands anyway.

haenoe commented:
Thanks for the quick reply!
I didn't try to compile after the rebase. Should be fixed now ^-^'
Quick question: it felt a bit weird adjusting the `Cargo.toml` to get access to the `rustc_middle::bug` macro. It seems the macro wasn't used in the `rustc_bultin_macros` before. Did I overlook something here?

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14190719803/job/39754636775) [(plain)](https://github.com/rust-lang/rust/commit/928f4e86c60d47b61cf17add7b19cdfbe58dede7/checks/39754636775/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.882 Building wheels for collected packages: reuse
#12 2.883   Building wheel for reuse (pyproject.toml): started
#12 3.099   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.100   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=5bb60f62728aaedff7162745ce743c7f2f55069b3e7f82e6a37d70df455797cc
#12 3.100   Stored in directory: /tmp/pip-ephem-wheel-cache-hvu_opqn/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.103 Successfully built reuse
#12 3.103 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.507 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.507 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.081 Collecting virtualenv
#12 4.133   Downloading virtualenv-20.30.0-py3-none-any.whl (4.3 MB)
#12 4.288      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 28.3 MB/s eta 0:00:00
#12 4.349 Collecting filelock<4,>=3.12.2
#12 4.357   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.396 Collecting platformdirs<5,>=3.9.1
#12 4.404   Downloading platformdirs-4.3.7-py3-none-any.whl (18 kB)
#12 4.426 Collecting distlib<1,>=0.3.7
#12 4.433   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.447      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 39.7 MB/s eta 0:00:00
#12 4.527 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.718 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.7 virtualenv-20.30.0
#12 4.719 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.8s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      143296 kB
DirectMap2M:     8245248 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.226
---
fmt check
Diff in /checkout/compiler/rustc_builtin_macros/src/autodiff.rs:20:
         PatKind, TyKind,
     };
     use rustc_expand::base::{Annotatable, ExtCtxt};
+    use rustc_middle::bug;
     use rustc_span::{Ident, Span, Symbol, kw, sym};
     use thin_vec::{ThinVec, thin_vec};
     use tracing::{debug, trace};
Diff in /checkout/compiler/rustc_builtin_macros/src/autodiff.rs:26:
-    use rustc_middle::bug;
 
     use crate::errors;
 
fmt: checked 5930 files
Build completed unsuccessfully in 0:01:20
```

</details>

oli-obk commented on compiler/rustc_builtin_macros/Cargo.toml:
```diff
@@ -21,6 +21,7 @@ rustc_index = { path = "../rustc_index" }
 rustc_lexer = { path = "../rustc_lexer" }
 rustc_lint_defs = { path = "../rustc_lint_defs" }
 rustc_macros = { path = "../rustc_macros" }
+rustc_middle = { path = "../rustc_middle" }
```
Yea, please don't add this dependency to `rustc_builtin_macros`. Crates before it in the crate graph can use `unreachable!` and friends

bors commented:
:umbrella: The latest upstream changes (presumably #138740) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

haenoe commented on compiler/rustc_builtin_macros/src/autodiff.rs:
```diff
@@ -145,27 +145,46 @@ mod llvm_enzyme {
             Annotatable::Item(iitem) => {
-                let (ident, sig) = match &iitem.kind {
-                    ItemKind::Fn(box ast::Fn { ident, sig, .. }) => (ident, sig),
+                let (sig, ident) = match &iitem.kind {
+                    ItemKind::Fn(box ast::Fn { sig, ident, .. }) => (sig, ident),
                     _ => {
                         dcx.emit_err(errors::AutoDiffInvalidApplication { span: item.span() });
                         return vec![item];
                     }
                 };
-                (*ident, sig.clone(), false)
+                (sig.clone(), iitem.vis.clone(), ident.clone())
```
I like the idea! However, I'm unsure how to replicate the current "wildcard" pattern logic in a good way using this approach.
Problem is that it `return`s from the scope of the outer `expand` function.
Currently can't think of a good solution, which doesn't re-introduce code duplication between the branches. But maybe I misunderstood you or you have a better idea! ^^'

oli-obk replied:
Well... all the code paths emit the same error, so we can change them to a `None` instead and report the error in a single site after the `match`

haenoe replied:
Yess! that is much better, thanks a lot

haenoe committed:
refactor: simplify function-info gathering

oli-obk commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 284672a9723f30eef6b9e54d360f9aa2254228c2 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 284672a9723f30eef6b9e54d360f9aa2254228c2 -->
<!-- homu: {"type":"Approved","sha":"284672a9723f30eef6b9e54d360f9aa2254228c2","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

ZuseZ4 commented:
Thank you two! But @oli-obk I'd be cautious with a rollup here, it will probably get a merge conflict in ~1hr, once my batching PR lands. I think github's merge algorithm likes to err on the site of caution.

oli-obk commented:
@bors r-

ZuseZ4 commented:
Unfortunate, but as predicted. @haenoe the diff should hopefully be easy to fix, can you rebase one more time, so we can add it back to the CI?

bors commented:
:umbrella: The latest upstream changes (presumably #139396) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

haenoe commented:
Yess of course! Will look at it when I'm at my computer in ~2 hours.
Glad to see things moving so quickly :D

haenoe commented:
This took longer than expected, because I ran into some very weird `LLVM ERROR`s and had to recompile a bunch of times xD

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14295990102/job/40063223305) [(plain)](https://github.com/rust-lang/rust/commit/0108d405fad58fb39a6ffda4d01a6b9313bdd4e2/checks/40063223305/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 19.9s done
#19 DONE 31.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error[E0433]: failed to resolve: use of undeclared type `ExprKind`
   --> compiler/rustc_builtin_macros/src/autodiff.rs:627:21
    |
627 |                     ExprKind::Path(Some(P(q)), ecx.path_ident(span, Ident::from_str("default")));
    |                     ^^^^^^^^ use of undeclared type `ExprKind`
    |
help: consider importing one of these enums
    |
7   +     use crate::autodiff::llvm_enzyme::ast::ExprKind;
    |
7   +     use rustc_ast::ExprKind;
    |
7   +     use rustc_hir::ExprKind;
    |
```

</details>

ZuseZ4 commented:
Yes, I also sometimes need to recompile llvm on updates, worst case I `rm -rf build`, but that's rarely necessary.

@bors r+ rollup

bors commented:
:pushpin: Commit bf69443a9f0fa9b44aaec36c1c470ad22a325c2a has been approved by `ZuseZ4`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ZuseZ4 bf69443a9f0fa9b44aaec36c1c470ad22a325c2a -->
<!-- homu: {"type":"Approved","sha":"bf69443a9f0fa9b44aaec36c1c470ad22a325c2a","approver":"ZuseZ4","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

ZuseZ4 commented:
My experience is that if you rebase across an LLVM update, then the very first `x` command will often work once,
and every following one (including x test) will cause the full llvm rebuild, presumably because it realizes that the llvm submodule has changed. 

But I also have to say that this specific rebase was strange, when I updated my local rust builds, two of them ran into an LLVM issue that some option was already known, so I had to delete the whole build folder, this happens rarely.

If you in the future run into an issue where x isn't rebuilding all of llvm, but everything starting from `rustc_llvm` onwards (without an obvious reason), then there is a good chance that rust-analyzer is causing it, feel free to ping me in that case.

But after all, the vast majority of rustc devs don't build LLVM at all, so there are also fewer people to test, report, and fix bootstrap issues related to it. I hope to get rid of the requirement of building LLVM from source, but that requires fighting cmake, something I am not too keen about.
